### DISCUSSION TITLE
Add Tiberian Pronunciation Tradition

### DISCUSSION THREAD

**Discussion by johnlockejrr** [charlesLoder/hebrew-transliteration#37] [Ideas]
It could be very useful to add The Tiberian Pronunciation Tradition (https://www.openbookpublishers.com/books/10.11647/obp.0163) as per Geoffrey Khan work. Not very hard, working on it myself for a personal project.

Example for Genesis 1:1

baʀ̟eːˈʃiːiθ bɔːˈʀ̟ɔː ʔɛloːˈhiːim ˈʔeːeθ haʃʃɔːˈmaːjim veˈʔeːeθ hɔːˈʔɔːʀ̟ɛsˁ

**Comment by charlesLoder** [comment]
@johnlockejrr 

That would be great!

I used Khan's work in the [syllabification package](https://github.com/charlesLoder/havarotjs) this package uses.

If you're able to do it, you can open a PR.

**Comment by johnlockejrr** [comment]
Yes, Tiberian Pronunciation Tradition has much to do with syllabification because that's how you know a vowel is long or short if is in an open syllable or a closed accentuated syllable. In Tiberian vocalization that's why they invented sheva, not to write down a "very short vowel" but to not allow a short vowel to be lenghtened if in an open syllable. Tiberian tradition noted vowels QUALITY and not QUANTITY etc.

Right now reading your code to better understant how to lenghten vowels when accentuated or in open syllables, remove quiescent Aleph (like in baʀ̟eːˈʃiːiθ) and add rules of sheva as normal short vowels with quality of a, ɔ, e, ɛ, o, u and i (khatefs don't have all the qualities).

**Comment by charlesLoder** [comment]
Let me know what kind of issues you run into. Some work may have to be done on the havarotjs package, and I have an idea to improve the `ADDITIONAL_FEATURES` property

**Reply by johnlockejrr** [reply]
Sure, I'll come back to you later with all the rules from Geoffrey Khan's work and the problems I encounter. Indeed, some changes should be made. One case is the שְׁתַּיִם word being transliterated as šǝtayim but in the Tiberian tradition had silent shewa (was the only exception in this case at a beginning of a word) being pronounced as [ˈʃtaːyim] or according to some medieval sources as [ʔɛʃˈtaːjim]. Or other cases when shewa was silent but is transliterated as vocal, e.g. שֹֽׁמְרִים֙ šōmǝrîm but according to Tiberian rules was pronounced [ˌʃoːmˈʀiːm] with silent shewa etc.

**Comment by charlesLoder** [comment]
>  One case is the שְׁתַּיִם word being transliterated as šǝtayim 

That one is tricky; I even mentioned it in the [syllabification docs](https://charlesloder.github.io/havarotjs/pages/Linguistic/syllabification.html#traditional).

>  transliterated as vocal, e.g. שֹֽׁמְרִים֙ šōmǝrîm 

The [`longVowel`](https://charlesloder.github.io/havarotjs/pages/Linguistic/syllabification.html#longvowles) option handles that.

When it is false, it is _šōmrîm_

**Comment by johnlockejrr** [comment]
About שְׁתַּיִם and other exceptions (in other cases) maybe they should be manually added to a list of exceptions?

**Comment by johnlockejrr** [comment]
Any idea where should I add some rules, let's say, for quality of shewa or vowel length (open syllable or accentuated closed syllable)? Delete quiescent Aleph, He, Waw, Yod etc. Can be added in schema? I already added a new tiberianKhan.js module with definitions of signs and sqnmlvy/wawShureq/STRESS_MARKER etc.
As an early test I get this:

**bǝʀ̟eʔˈʃîθ bɔˈʀ̟ɔʔ ʔĕloˈhîm ˈʔeθ haʃʃɔˈmajim vǝˈʔeθ hɔʔɔʀ̟ɛsˁ**

I need to get this:

**baʀ̟eːˈʃiːiθ bɔːˈʀ̟ɔː ʔɛloːˈhiːim ˈʔeːeθ haʃʃɔːˈmaːjim veˈʔeːeθ hɔːˈʔɔːʀ̟ɛsˁ**

**Comment by charlesLoder** [comment]
@johnlockejrr ,

Check out this [new branch](https://github.com/charlesLoder/hebrew-transliteration/tree/add-feature-cb) where  you can use a callback for an additional feature. See the test for an example.

If you're working on a branch, you can rebase on that branch.

**Comment by johnlockejrr** [comment]
Had some luck with `ADDITIONAL_FEATURES` but I don't have much documentation for it. Can be `TRANSLITERATION` used like in REGEX something like search for (HEBREW) `^([\u05B0])(\u05D0\u05B8)` and replace with (TRANSLITERATION) `ɔ\2`? So I can make only one rule for all vocal shevas before gutturals (אהחע) and yod (י)? What I did for now is different rules for all that I'm not happy with, like:

1. { FEATURE: "word", HEBREW: "\u05B0\u05D4\u05B4", TRANSLITERATION: "ihi" },
2. { FEATURE: "word", HEBREW: "\u05B0\u05D4\u05B5", TRANSLITERATION: "ehe" },
3. { FEATURE: "word", HEBREW: "\u05B0\u05D4\u05B6", TRANSLITERATION: "ɛhɛ" },
4. { FEATURE: "word", HEBREW: "\u05B0\u05D4\u05B7", TRANSLITERATION: "aha" },
5. { FEATURE: "word", HEBREW: "\u05B0\u05D4\u05B8", TRANSLITERATION: "ɔhɔ" },
6. { FEATURE: "word", HEBREW: "\u05B0\u05D4\u05B9", TRANSLITERATION: "oho" },
7. { FEATURE: "word", HEBREW: "\u05B0\u05D4\u05BB", TRANSLITERATION: "uhu" }

**Comment by charlesLoder** [comment]
> I don't have much documentation for it.

Yeah, documentation hasn't been my strong suit.

Let me fiddle around with the callback to see it I can get something

**Comment by charlesLoder** [comment]
It's a little rough

```js
const heb = require("./dist/index");
const additionalFeatureTransliteration = require("./dist/rules").additionalFeatureTransliteration;

const str = "יְהִי";
const result = heb.transliterate(str, {
  ADDITIONAL_FEATURES: [
    {
      FEATURE: "cluster",
      HEBREW: "\u05B0",
      TRANSLITERATION: (cluster, transliteration, schema) => {
        const shewa = new RegExp(transliteration, "u");
        const clusterText = cluster.text;
        /**
         * @type {Cluster}
         */
        const next = cluster.next;
        const gutturalYodVowel = /[אהחעי]([\u{05B1}-\u{05BB}\u{05C7}])/u;
        const match = next.text.match(gutturalYodVowel);
        if (shewa.test(clusterText) && match) {
          return additionalFeatureTransliteration(clusterText, shewa, match[1], schema);
        }
        return clusterText;
      }
    }
  ]
});

console.log(result);

// yihî
```

Some of the naming conventions need to be improved in the callback.

**Comment by johnlockejrr** [comment]
Wonderful!! I will give it a go.

**Comment by johnlockejrr** [comment]
Odd enough I get this result on the console:

```
y(cluster, transliteration, schema) => {
        const shewa = new RegExp(transliteration, "u");
        const clusterText = cluster.text;
        /**
         * @type {Cluster}
         */
        const next = cluster.next;
        const gutturalYodVowel = /[ʾhḥʿy]([\u{05B1}-\u{05BB}\u{05C7}])/u;
        const match = next.text.match(gutturalYodVowel);
        if (shewa.test(clusterText) && match) {
          return additionalFeatureTransliteration(clusterText, shewa, match[1], schema);
        }
        return clusterText;
      }hî
```

Is verbose or smth?

**Reply by charlesLoder** [reply]
This should give the correct result

```diff
y(cluster, transliteration, schema) => {
        const shewa = new RegExp(transliteration, "u");
        const clusterText = cluster.text;
        /**
         * @type {Cluster}
         */
        const next = cluster.next;
-        const gutturalYodVowel = /[ʾhḥʿy]([\u{05B1}-\u{05BB}\u{05C7}])/u;
+        const gutturalYodVowel = /[אהחעי]([\u{05B1}-\u{05BB}\u{05C7}])/u;
        const match = next.text.match(gutturalYodVowel);
        if (shewa.test(clusterText) && match) {
          return additionalFeatureTransliteration(clusterText, shewa, match[1], schema);
        }
        return clusterText;
      }
```

**Comment by johnlockejrr** [comment]
That wasn't an issue, in the JS file has `const gutturalYodVowel = /[אהחעי]([\u{05B1}-\u{05BB}\u{05C7}])/u;` but the output on the console is `const gutturalYodVowel = /[ʾhḥʿy]([\u{05B1}-\u{05BB}\u{05C7}])/u;`

```
john@vps:~/hebrew-transliteration$ cat test.js
const heb = require("./dist/index");
const additionalFeatureTransliteration = require("./dist/rules").additionalFeatureTransliteration;

const str = "וְאֵ֥ת";
const result = heb.transliterate(str, {
  ADDITIONAL_FEATURES: [
    {
      FEATURE: "cluster",
      HEBREW: "\u05B0",
      TRANSLITERATION: (cluster, transliteration, schema) => {
        const shewa = new RegExp(transliteration, "u");
        const clusterText = cluster.text;
        /**
         * @type {Cluster}
         */
        const next = cluster.next;
        const gutturalYodVowel = /[אהחעי]([\u{05B1}-\u{05BB}\u{05C7}])/u;
        const match = next.text.match(gutturalYodVowel);
        if (shewa.test(clusterText) && match) {
          return additionalFeatureTransliteration(clusterText, shewa, match[1], schema);
        }
        return clusterText;
      }
    }
  ]
});

console.log(result);
john@vps:~/hebrew-transliteration$ node test.js
w(cluster, transliteration, schema) => {
        const shewa = new RegExp(transliteration, "u");
        const clusterText = cluster.text;
        /**
         * @type {Cluster}
         */
        const next = cluster.next;
        const gutturalYodVowel = /[ʾhḥʿy]([\u{05B1}-\u{05BB}\u{05C7}])/u;
        const match = next.text.match(gutturalYodVowel);
        if (shewa.test(clusterText) && match) {
          return additionalFeatureTransliteration(clusterText, shewa, match[1], schema);
        }
        return clusterText;
      }ʾēt
john@vps:~/hebrew-transliteration$
```

**Comment by charlesLoder** [comment]
Ok, I’ll take a look after the holidays

**Comment by johnlockejrr** [comment]
Thanks @charlesLoder, happy holidays!

**Comment by charlesLoder** [comment]
It looks like the build command wasn't run after checking out the latest branch.

If you're still on `main`, checkokut the latest branch and build: 

```
git switch add-feature-cb
npm run build
node test.js
```

**Comment by johnlockejrr** [comment]
***@***.***:~/hebrew-transliteration$ npm run build

> ***@***.*** build
> tsc

***@***.***:~/hebrew-transliteration$ node test.js
w(cluster, transliteration, schema) => {
        const shewa = new RegExp(transliteration, "u");
        const clusterText = cluster.text;
        /**
         * @type {Cluster}
         */
        const next = cluster.next;
        const gutturalYodVowel = /[ʾhḥʿy]([\u{05B1}-\u{05BB}\u{05C7}])/u;
        const match = next.text.match(gutturalYodVowel);
        if (shewa.test(clusterText) && match) {
          return additionalFeatureTransliteration(clusterText, shewa,
match[1], schema);
        }
        return clusterText;
      }ʾēt

**Comment by johnlockejrr** [comment]
This is the code in my test.js, does it run for you? For me is just verbose and outputs the js code:

```
const heb = require("./dist/index");
const additionalFeatureTransliteration = require("./dist/rules").additionalFeatureTransliteration;

const str = "וְאֵ֥ת";
const result = heb.transliterate(str, {
  ADDITIONAL_FEATURES: [
    {
      FEATURE: "cluster",
      HEBREW: "\u05B0",
      TRANSLITERATION: (cluster, transliteration, schema) => {
        const shewa = new RegExp(transliteration, "u");
        const clusterText = cluster.text;
        /**
         * @type {Cluster}
         */
        const next = cluster.next;
        const gutturalYodVowel = /[אהחעי]([\u{05B1}-\u{05BB}\u{05C7}])/u;
        const match = next.text.match(gutturalYodVowel);
        if (shewa.test(clusterText) && match) {
          return additionalFeatureTransliteration(clusterText, shewa, match[1], schema);
        }
        return clusterText;
      }
    }
  ]
});

console.log(result);
```

**Comment by charlesLoder** [comment]
@johnlockejrr ,

When I run your code on the [`add-feature-cb`](https://github.com/charlesLoder/hebrew-transliteration/tree/add-feature-cb) branch, I get the result of
```
wēʾēt
```


When I run it on the `main` branch, I get 

```
w(cluster, transliteration, schema) => {
        const shewa = new RegExp(transliteration, "u");
        const clusterText = cluster.text;
        /**
         * @type {Cluster}
         */
        const next = cluster.next;
        const gutturalYodVowel = /[ʾhḥʿy]([\u{05B1}-\u{05BB}\u{05C7}])/u;
        const match = next.text.match(gutturalYodVowel);
        if (shewa.test(clusterText) && match) {
          return additionalFeatureTransliteration(clusterText, shewa, match[1], schema);
        }
        return clusterText;
      }ʾēt
```

**Comment by johnlockejrr** [comment]
Now that is very very strange... I will try on another system, is just weird. Here is my full log:

```
john@vps:~$ git clone https://github.com/charlesLoder/hebrew-transliteration
Cloning into 'hebrew-transliteration'...
remote: Enumerating objects: 1447, done.
remote: Counting objects: 100% (530/530), done.
remote: Compressing objects: 100% (251/251), done.
remote: Total 1447 (delta 341), reused 437 (delta 275), pack-reused 917
Receiving objects: 100% (1447/1447), 1.28 MiB | 8.57 MiB/s, done.
Resolving deltas: 100% (971/971), done.
john@vps:~$ cd hebrew-transliteration
john@vps:~/hebrew-transliteration$ git checkout -b add-feature-cb
Switched to a new branch 'add-feature-cb'
john@vps:~/hebrew-transliteration$ npm install
npm WARN deprecated @npmcli/move-file@2.0.1: This functionality has been moved to @npmcli/fs

> hebrew-transliteration@2.3.1 prepare
> npm run build


> hebrew-transliteration@2.3.1 build
> tsc


added 694 packages, and audited 695 packages in 16s

101 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
john@vps:~/hebrew-transliteration$ npm run build

> hebrew-transliteration@2.3.1 build
> tsc

john@vps:~/hebrew-transliteration$ node test.js
w(cluster, transliteration, schema) => {
        const shewa = new RegExp(transliteration, "u");
        const clusterText = cluster.text;
        /**
         * @type {Cluster}
         */
        const next = cluster.next;
        const gutturalYodVowel = /[ʾhḥʿy]([\u{05B1}-\u{05BB}\u{05C7}])/u;
        const match = next.text.match(gutturalYodVowel);
        if (shewa.test(clusterText) && match) {
          return additionalFeatureTransliteration(clusterText, shewa, match[1], schema);
        }
        return clusterText;
      }ʾēt

```

**Comment by johnlockejrr** [comment]
Tried it on another machine, same result, I'm on `add-feature-cb` branch.

<img width="362" alt="Linux Mint 19 64bit-2022-12-27-20-27-56" src="https://user-images.githubusercontent.com/16368414/209713757-698915f4-c06f-48b5-8cf8-3bd158f9e43b.png">

**Comment by johnlockejrr** [comment]
Somehow, even being on `add-feature-cb` my git wasn't cloning that branch, so I cloned dirrectly that branch without the main one `git clone --branch add-feature-cb https://github.com/charlesLoder/hebrew-transliteration.git`
Now seems to work.

**Comment by johnlockejrr** [comment]
Good, now works, kind of... :)
Getting weird transliterations, like disappearing dagesh (^b -> ^v), sin transliterated as shin, silent shewa made vocal etc.

`ˈvajiˈħi: jaʕaˈq̟ov vɛˈʔɛʀ̟ɛsˁ misˁǝˈʀ̟ajim ʃׁǝˈvaʕ ʕɛʃׂǝˈʀ̟ɛ: ʃɔˈnɔ: vajiˈhi: jǝme:-jaʕaˈq̟ov ʃׁǝˈne: ħaˈjɔw ˈʃɛvaʕ ʃɔˈni:m vaʔaʀ̟ǝbɔˈʕi:m u:maˈʔaθ ʃɔnɔ:`

should have been (not yet Tiberian! working on it):

`ˈvajiˈħi: jaʕaˈq̟ov bɛˈʔɛʀ̟ɛsˁ misˁˈʀ̟ajim ʃǝˈvaʕ ʕɛʃˈʀ̟ɛ: ʃɔˈnɔ: vajiˈhi: jǝme:-jaʕaˈq̟ov ʃˈne: ħaˈjɔw ˈʃɛvaʕ ʃɔˈni:m vaʔaʀ̟bɔˈʕi:m u:maˈʔaθ ʃɔnɔ:`

Using SBL schema I get this:

`wayiḥî yaʿăqōb beʾereṣ miṣǝrayim šׁǝbaʿ ʿešׂǝrê šānâ wayihî yǝmê-yaʿăqōb šׁǝnê ḥayāyw šebaʿ šānîm waʾarǝbāʿîm ûmaʾat šānâ`

Heb.

`וַ֯יְחִ֤י יַֽעֲקֹב֙ בְּאֶ֣רֶץ מִצְרַ֔יִם שְׁבַ֥ע עֶשְׂרֵ֖ה שָׁנָ֑ה וַיְהִ֤י יְמֵי־יַֽעֲקֹב֙ שְׁנֵ֣י חַיָּ֔יו שֶׁ֣בַע שָׁנִ֔ים וְאַרְבָּעִ֥ים וּמְאַ֖ת שָׁנָֽה`

**Comment by charlesLoder** [comment]
@johnlockejrr 

Yeah, right now everything is a little too tightly coupled.

Because Tiberian moves more into the realm of transcription, rather than transliteration, there are still a lot of improvements that need to be made

**Comment by charlesLoder** [comment]
fyi, I recently had a new addition to my family so work on this project has been slowed. I'll keep working on havarotjs which help improve this, but it will be a while before I get any solid work done on this.
